// umami-api/dist/UmamiAPIClient.js
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const axios_1 = __importDefault(require("axios"));
const time_periods_1 = require("./utils/time-periods");
const DEFAULT_USER_AGENT = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:102.0) Gecko/20100101 Firefox/102.0";
function _richError(message, cause, options) {
    if (!options)
        options = "None specified";
    options = JSON.stringify(options);
    if (!cause)
        cause = "None specified";
    cause = cause.toString();
    return new Error(`${message}\nOptions: ${options}\nStacktrace:`, { cause });
}
/**
 * Umami API Client
 */
class UmamiAPIClient {
    /**
     * @param server The Umami installation hostname (e.g. app.umami.is). The protocol, if present, will be removed.
     * @param username Username of the user you want to login
     * @param password Password of the user you want to login
     * @returns An authenticated class instance
     * @see {@link https://github.com/umami-software/umami/blob/master/pages/api/auth/login.js Relevant Umami source code}
     */
    constructor(server, username, password) {
        this._lastAuthCheck = Date.now();
        this._defaultPeriod = "24h";
        this._defaultUnit = "hour";
        this._defaultTZ = "America/Toronto";
        this._defaultMetricType = "url";
        this._defaultUserAgent = DEFAULT_USER_AGENT;
        if (!server)
            throw new Error("A server hostname is required");
        server = server.replace(/https?:\/\//, "").replace(/\/$/, "");
        if (!username || !password)
            throw new Error("A username and a password are required");
        this._axios = axios_1.default.create({
            baseURL: `https://${server}/api`,
            timeout: 1000,
        });
        this._axios.interceptors.request.use(this._verifyAuth.bind(this));
        this._auth = this._axios.post("/auth/login", { username, password }).catch((error) => {
            throw _richError("Login failed", error, { server, username });
        });
    }
    setDefaultPeriod(period) {
        this._defaultPeriod = period;
    }
    setDefaultUnit(unit) {
        this._defaultUnit = unit;
    }
    setDefaultTZ(tz) {
        this._defaultTZ = tz;
    }
    setDefaultMetricType(metricType) {
        this._defaultMetricType = metricType;
    }
    setDefaultUserAgent(userAgent) {
        this._defaultUserAgent = userAgent;
    }
    async getCurrentUser() {
        return (await this._auth).data.user;
    }
    async _verifyAuth(config) {
        if (config.url == "/auth/login" || config.url == "/collect")
            return config;
        const auth = await this._auth;
        config.headers = { ...config.headers, Authorization: `Bearer ${auth.data.token}` };
        if (config.url == "/auth/verify")
            return config;
        if (this._lastAuthCheck + 60 * 60 * 1000 < Date.now()) {
            this._lastAuthCheck = Date.now();
            try {
                await this._axios.get("/auth/verify");
            }
            catch (error) {
                throw _richError("Could not verify authentication", error, { axiosConfig: config });
            }
        }
        return config;
    }
    async collect(type, payload, userAgent = this._defaultUserAgent) {
        try {
            if (!userAgent)
                throw new Error("A user agent is required. See https://umami.is/docs/api");
            const { data } = await this._axios.post("/collect", { type, payload }, { headers: { "User-Agent": userAgent } });
            return data;
        }
        catch (error) {
            throw _richError("Could not collect", error, { type, payload, userAgent });
        }
    }
    static async collect(server, type, payload, userAgent = DEFAULT_USER_AGENT) {
        server = server.replace(/https?:\/\//, "").replace(/\/$/, "");
        try {
            if (!userAgent)
                throw new Error("A user agent is required. See https://umami.is/docs/api");
            const { data } = await axios_1.default.post(`https://${server}/api/collect`, { type, payload }, { headers: { "User-Agent": userAgent } });
            return data;
        }
        catch (error) {
            throw _richError("Could not collect", error, { type, payload, userAgent });
        }
    }
    /**
     * Creates a new website and returns its information.
     * @param options.domain The domain name of the website (e.g. umami.is)
     * @param options.name The name of the website (usually the same as the domain)
     * @param options.enable_share_url Whether or not to enable public sharing.
     * @returns The new website's information
     * @see {@link https://github.com/umami-software/umami/blob/master/pages/api/website/index.js#L36 Relevant Umami source code}
     */
    async createWebsite(options) {
        try {
            const { data } = await this._axios.post("/website", options);
            return data;
        }
        catch (error) {
            throw _richError("Could not create website", error, { options });
        }
    }
    /**
     * Updates a website and returns its information.
     * @param website_id The website's ID (not UUID)
     * @param options.domain The domain name of the website (e.g. umami.is)
     * @param options.name The name of the website (usually the same as the domain)
     * @param options.enable_share_url Whether or not to enable public sharing.
     * @returns The website's information
     * @see {@link https://github.com/umami-software/umami/blob/master/pages/api/website/index.js#L30 Relevant Umami source code}
     */
    async updateWebsite(website_id, options) {
        try {
            const { data } = await this._axios.post("/website", { website_id, ...options });
            return data;
        }
        catch (error) {
            throw _richError("Could not update website", error, { website_id, options });
        }
    }
    async getWebsite(website_id = null) {
        try {
            if (website_id == null) {
                const websites = await this.getWebsites();
                return websites[0];
            }
            const { data } = await this._axios.get(`/website/${website_id}`);
            return data;
        }
        catch (error) {
            throw _richError("Could not get website", error, { website_id });
        }
    }
    /**
     * Get a website by a property
     * @param key The property to check
     * @param value The value to check the property against
     * @returns The website
     * @see {@link https://github.com/umami-software/umami/blob/master/pages/api/websites/index.js Relevant Umami source code}
     *
     * @example
     * Get a website by domain name
     * ```ts
     * const website = await instance.getWebsiteBy("domain", "example.com");
     * ```
     */
    async getWebsiteBy(key, value) {
        if (key == "share_id") {
            try {
                const { data } = await this._axios.get(`/share/${value}`);
                return await this.getWebsite(data.websiteId);
            }
            catch (error) {
                throw _richError("Could not find website", error, { key, value });
            }
        }
        if (key == "website_id") {
            try {
                return await this.getWebsite(value);
            }
            catch (error) {
                throw _richError("Could not find website", error, { key, value });
            }
        }
        const websites = await this.getWebsites();
        const website = websites.find((website) => website[key] == value);
        if (!website) {
            throw _richError("Could not find website", null, { key, value });
        }
        return website;
    }
    /**
     * Resets a website's stats
     * @param website_id The website's ID (not UUID)
     * @see {@link https://github.com/umami-software/umami/blob/master/pages/api/website/[id]/reset.js Relevant Umami source code}
     */
    async resetWebsite(website_id) {
        try {
            const { data } = await this._axios.post(`/website/${website_id}/reset`);
            return data;
        }
        catch (error) {
            throw _richError("Could not reset website", error, { website_id });
        }
    }
    /**
     * Delete a website
     * @param website_id The website's ID (not UUID)
     * @see {@link https://github.com/umami-software/umami/blob/master/pages/api/website/[id]/index.js Relevant Umami source code}
     */
    async deleteWebsite(website_id) {
        try {
            await this._axios.delete(`/website/${website_id}`);
        }
        catch (error) {
            throw _richError("Could not delete website", error, { website_id });
        }
    }
    /**
     * Gets tracked websites
     * @param options.include_all Whether or not to include all websites (admin only)
     * @param options.user_id The user to query websites from (admin only, if not your own user id)
     * @returns An array of tracked websites
     * @see {@link https://github.com/umami-software/umami/blob/master/pages/api/websites/index.js Relevant Umami source code}
     */
    async getWebsites(options) {
        try {
            const { data } = await this._axios.get("/websites", { params: options });
            return data;
        }
        catch (error) {
            throw _richError("Could not get websites", error, { options });
        }
    }
    /**
     * Gets the stats of a website from a specified time period using it's ID
     * @param website_id The website's ID (not UUID)
     * @param options.period The time period of stats to return
     * @param options.url Filter stats by URL
     * @param options.referrer Filter stats by referrer
     * @param options.os Filter stats by OS
     * @param options.browser Filter stats by browser
     * @param options.device Filter stats by device
     * @param options.country Filter stats by country
     * @returns The website's stats from the specified time period
     * @see {@link https://github.com/umami-software/umami/blob/master/pages/api/website/[id]/stats.js Relevant Umami source code}
     */
    async getStats(website_id, options) {
        var _a;
        const { start_at, end_at } = (0, time_periods_1.convertPeriodToTime)((_a = options.period) !== null && _a !== void 0 ? _a : this._defaultPeriod);
        const params = { ...options, start_at, end_at };
        try {
            const { data } = await this._axios.get(`/website/${website_id}/stats`, { params });
            return data;
        }
        catch (error) {
            throw _richError("Could not get stats", error, { website_id, params });
        }
    }
    /**
     * Gets the pageviews of a website from a specified time period using it's ID
     * @param website_id The website's ID (not UUID)
     * @param options.period The time period of pageviews to return
     * @param options.unit The interval of time/precision of the returned pageviews
     * @param options.tz The timezone you're in (defaults to "America/Toronto")
     * @param options.url Filter pageviews by URL
     * @param options.referrer Filter pageviews by referrer
     * @param options.os Filter pageviews by OS
     * @param options.browser Filter pageviews by browser
     * @param options.device Filter pageviews by device
     * @param options.country Filter pageviews by country
     * @returns The website's pageviews from the specified time period
     * @see {@link https://github.com/umami-software/umami/blob/master/pages/api/website/[id]/pageviews.js Relevant Umami source code}
     */
    async getPageviews(website_id, options) {
        var _a, _b, _c;
        const { start_at, end_at } = (0, time_periods_1.convertPeriodToTime)((_a = options.period) !== null && _a !== void 0 ? _a : this._defaultPeriod);
        const unit = (_b = options.unit) !== null && _b !== void 0 ? _b : this._defaultUnit;
        const tz = (_c = options.tz) !== null && _c !== void 0 ? _c : this._defaultTZ;
        const params = { ...options, start_at, end_at, unit, tz };
        try {
            const { data } = await this._axios.get(`/website/${website_id}/pageviews`, { params });
            return data;
        }
        catch (error) {
            throw _richError("Could not get pageviews", error, { website_id, params });
        }
    }
    /**
     * Gets the events of a website from a specified time period using it's ID
     * @param website_id The website's ID (not UUID)
     * @param options.period The time period of events to return
     * @param options.unit The interval of time/precision of the returned events
     * @param options.tz The timezone you're in (defaults to "America/Toronto")
     * @param options.url The url where the event happened.
     * @param options.event_type The type of event to request.
     * @returns An array of events from the specified time period
     * @see {@link https://github.com/umami-software/umami/blob/master/pages/api/website/[id]/events.js Relevant Umami source code}
     */
    async getEvents(website_id, options) {
        var _a, _b, _c;
        const { start_at, end_at } = (0, time_periods_1.convertPeriodToTime)((_a = options.period) !== null && _a !== void 0 ? _a : this._defaultPeriod);
        const unit = (_b = options.unit) !== null && _b !== void 0 ? _b : this._defaultUnit;
        const tz = (_c = options.tz) !== null && _c !== void 0 ? _c : this._defaultTZ;
        const params = { start_at, end_at, unit, tz, url: options.url, event_type: options.event_type };
        try {
            const { data } = await this._axios.get(`/website/${website_id}/events`, { params });
            return data;
        }
        catch (error) {
            throw _richError("Could not get events", error, { website_id, params });
        }
    }
    /**
     * Gets the total number of events by a filter
     * @param website_id The website's ID (not UUID)
     * @param options.filter The field to filter by
     * @param options.value The value to match the field against
     * @param options.period The time period of events to return
     * @returns The total number of events matching the filter
     * @see {@link https://github.com/umami-software/umami/blob/master/pages/api/websites/index.js Relevant Umami source code}
     */
    async getEventsByName(website_id, options) {
        try {
            const events = await this.getEvents(website_id, {
                period: options.period,
                unit: options.unit,
                tz: options.tz,
                url: options.url,
                event_type: options.event_type,
            });
            return events.filter((event) => event.x == options.name);
        }
        catch (error) {
            throw _richError("Could not get events by name", error, { website_id, options });
        }
    }
    /**
     * Gets a type of metrics of a website from a specified time period using it's ID
     * @param website_id The website's ID (not UUID)
     * @param options.period The time period of events to return
     * @param options.type The type of metric to get. Defaults to url
     * @param options.url Filter metrics by URL
     * @param options.referrer Filter metrics by referrer
     * @param options.os Filter metrics by OS
     * @param options.browser Filter metrics by browser
     * @param options.device Filter metrics by device
     * @param options.country Filter metrics by country
     * @returns An array of metrics from the specified time period
     * @see {@link https://github.com/umami-software/umami/blob/master/pages/api/website/[id]/metrics.js Relevant Umami source code}
     */
    async getMetrics(website_id, options) {
        var _a, _b;
        const { start_at, end_at } = (0, time_periods_1.convertPeriodToTime)((_a = options.period) !== null && _a !== void 0 ? _a : this._defaultPeriod);
        const type = (_b = options.type) !== null && _b !== void 0 ? _b : this._defaultMetricType;
        const params = { ...options, start_at, end_at, type };
        try {
            const { data } = await this._axios.get(`/website/${website_id}/metrics`, { params });
            return data;
        }
        catch (error) {
            throw _richError("Could not get metrics", error, { website_id, params });
        }
    }
    /**
     * Gets the active visitors of a website
     * @param website_id The website's ID (not UUID)
     * @returns
     * @see {@link https://github.com/umami-software/umami/blob/master/pages/api/website/[id]/active.js Relevant Umami source code}
     */
    async getActiveVisitors(website_id) {
        try {
            const { data } = await this._axios.get(`/website/${website_id}/active`);
            return data;
        }
        catch (error) {
            throw _richError("Could not get active visitors", error, { website_id });
        }
    }
    /*** ADMIN ONLY FUNCTIONS ***/
    /**
     * Creates a user account (admin only)
     * @param options.username The username
     * @param options.password The password
     * @returns The user account
     * @see {@link https://github.com/umami-software/umami/blob/master/pages/api/account/index.js Relevant Umami source code}
     */
    async createAccount(options) {
        try {
            const { data } = await this._axios.post("/account", options);
            return data;
        }
        catch (error) {
            throw _richError("Could not create account", error, { options });
        }
    }
    /**
     * Updates a user account
     * @param user_id User ID to update
     * @param options.username New username (admin only)
     * @param options.password New password
     * @param options.is_admin New admin status (admin only)
     * @returns The user account
     * @see {@link https://github.com/umami-software/umami/blob/master/pages/api/account/index.js Relevant Umami source code}
     */
    async updateAccount(user_id, options) {
        try {
            const { data } = await this._axios.post("/account", { user_id, ...options });
            return data;
        }
        catch (error) {
            throw _richError("Could not update account", error, { user_id, options });
        }
    }
    /**
     * Updates a user account password
     * @param user_id User ID to update
     * @param options.current_password Current password
     * @param options.new_password New password
     * @returns The user account
     * @see {@link https://github.com/umami-software/umami/blob/master/pages/api/account/password.js Relevant Umami source code}
     */
    async changePassword(user_id, options) {
        try {
            const { data } = await this._axios.post("/account", { user_id, ...options });
            return data;
        }
        catch (error) {
            throw _richError("Could not update password", error, { user_id, options });
        }
    }
    /**
     * Gets all the user accounts (admin only)
     * @returns An array of all the user accounts
     * @see {@link https://github.com/umami-software/umami/blob/master/pages/api/accounts/index.js Relevant Umami source code}
     */
    async getAccounts() {
        try {
            const { data } = await this._axios.get("/accounts");
            return data;
        }
        catch (error) {
            throw _richError("Could not get accounts", error);
        }
    }
    /**
     * Gets a user account (admin only)
     * @param user_id The user ID
     * @returns The user account
     * @see {@link https://github.com/umami-software/umami/blob/master/pages/api/account/[id].js Relevant Umami source code}
     */
    async getAccount(user_id) {
        try {
            const { data } = await this._axios.get(`/account/${user_id}`);
            return data;
        }
        catch (error) {
            throw _richError("Could not get account", error, { user_id });
        }
    }
    /**
     * Deletes a user account (admin only)
     * @param user_id The user ID
     * @see {@link https://github.com/umami-software/umami/blob/master/pages/api/account/[id].js Relevant Umami source code}
     */
    async deleteAccount(user_id) {
        try {
            await this._axios.delete(`/account/${user_id}`);
        }
        catch (error) {
            throw _richError("Could not get account", error, { user_id });
        }
    }
}
exports.default = UmamiAPIClient;
