// umami-api/dist/UmamiAPI.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const axios_1 = __importDefault(require("axios"));
const time_periods_1 = require("./utils/time-periods");
class BaseUmamiAPI {
    /**
     * @param server The Umami installation hostname (e.g. app.umami.is). The protocol, if present, will be removed.
     */
    constructor(server) {
        if (!server) {
            throw new Error("A server hostname is required");
        }
        this._server = server;
        this._axios = axios_1.default.create({
            baseURL: `https://${this._server.replace(/https?:\/\//, "")}/api`,
            timeout: 1000,
        });
    }
    async collect(type, payload, userAgent = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:102.0) Gecko/20100101 Firefox/102.0") {
        if (userAgent === "") {
            throw new Error("A user agent is required for the /api/collect endpoint to work. See https://umami.is/docs/api");
        }
        try {
            const { data } = await this._axios.post("/collect", { type, payload }, { headers: { "User-Agent": userAgent } });
            return data;
        }
        catch (error) {
            throw new Error(`Collect failed`, { cause: error.toString() });
        }
    }
}
class UmamiAPI extends BaseUmamiAPI {
    /**
     * Authenticates a user, to be able to use authenticated API functions
     * @param username Username of the user you want to login
     * @param password Password of the user you want to login
     * @returns An authenticated class instance, able to use more functionality.
     * @see {@link https://github.com/umami-software/umami/blob/master/pages/api/auth/login.js Relevant Umami source code}
     */
    async auth(username, password) {
        try {
            const { data } = await this._axios.post("/auth/login", {
                username,
                password,
            });
            return new AuthenticatedUmamiAPI(this._server, data);
        }
        catch (error) {
            throw new Error("Login failed", { cause: error.toString() });
        }
    }
}
exports.default = UmamiAPI;
class AuthenticatedUmamiAPI extends BaseUmamiAPI {
    constructor(server, auth) {
        super(server);
        this._defaultPeriod = "24h";
        this._auth = auth;
        this._axios.defaults.headers.common["Authorization"] = `Bearer ${this._auth.token}`;
        this._lastCheck = Date.now();
    }
    setDefaultPeriod(period) {
        this._defaultPeriod = period;
    }
    getDefaultPeriod() {
        return this._defaultPeriod;
    }
    getCurrentUser() {
        return this._auth.user;
    }
    assertAdmin(message) {
        if (!this._auth.user.is_admin) {
            throw new Error(message);
        }
    }
    async verifyAuth() {
        if (this._lastCheck + 60 * 60 * 1000 < Date.now()) {
            try {
                const { data } = await this._axios.get("/auth/verify");
                this._auth.user = data;
            }
            catch (error) {
                throw new Error("Could not verify authentication", { cause: error.toString() });
            }
        }
    }
    async POST(endpoint, payload) {
        await this.verifyAuth();
        try {
            const { data } = await this._axios.post(endpoint, payload);
            return data;
        }
        catch (error) {
            throw new Error(`POST request to /api${endpoint} failed`, { cause: error.toString() });
        }
    }
    async GET(endpoint, params) {
        await this.verifyAuth();
        try {
            const { data } = await this._axios.get(endpoint, { params });
            return data;
        }
        catch (error) {
            throw new Error(`GET request to /api${endpoint} failed with params: ${params.toString()}`, {
                cause: error.toString(),
            });
        }
    }
    async DELETE(endpoint) {
        await this.verifyAuth();
        try {
            const { data } = await this._axios.delete(endpoint);
            return data;
        }
        catch (error) {
            throw new Error(`DELETE request to /api${endpoint} failed`, { cause: error.toString() });
        }
    }
    /**
     * Creates a new website and returns its information.
     * @param options.domain The domain name of the website (e.g. umami.is)
     * @param options.name The name of the website (usually the same as the domain)
     * @param options.enable_share_url Whether or not to enable public sharing.
     * @returns The new website's information
     * @see {@link https://github.com/umami-software/umami/blob/master/pages/api/website/index.js#L36 Relevant Umami source code}
     */
    async createWebsite(options) {
        return await this.POST("/website", options);
    }
    /**
     * Updates a website and returns its information.
     * @param website_id The website's ID (not UUID)
     * @param options.domain The domain name of the website (e.g. umami.is)
     * @param options.name The name of the website (usually the same as the domain)
     * @param options.enable_share_url Whether or not to enable public sharing.
     * @returns The website's information
     * @see {@link https://github.com/umami-software/umami/blob/master/pages/api/website/index.js#L30 Relevant Umami source code}
     */
    async updateWebsite(website_id, options) {
        return await this.POST("/website", { website_id, ...options });
    }
    async getWebsite(keyOrId, value) {
        if ((!keyOrId && !value) || (keyOrId && value)) {
            const websites = await this.getWebsites();
            if (!keyOrId) {
                return websites[0];
            }
            return websites.find((website) => website[keyOrId] == value);
        }
        return await this.GET(`/website/${keyOrId}`);
    }
    async resetWebsite(website_id) {
        return await this.GET(`/website/${website_id}/reset`);
    }
    async deleteWebsite(website_id) {
        return await this.DELETE(`/website/${website_id}`);
    }
    /**
     * Gets tracked websites
     * @param options.include_all Whether or not to include all websites (admin only)
     * @param options.user_id The user to query websites from (admin only, if not your own user id)
     * @returns An array of tracked websites
     * @see {@link https://github.com/umami-software/umami/blob/master/pages/api/websites/index.js Relevant Umami source code}
     */
    async getWebsites(options) {
        if (options.include_all || options.user_id) {
            this.assertAdmin("You are not authorized to view these websites");
        }
        return await this.GET("/websites", options);
    }
    /**
     * Gets the stats of a website from a specified time period using it's ID
     * @param website_id The website's ID (not UUID)
     * @param period The time period of stats to return
     * @returns The website's stats from the specified time period
     * @see {@link https://github.com/umami-software/umami/blob/master/pages/api/website/[id]/stats.js Relevant Umami source code}
     */
    async getStats(website_id, period = this._defaultPeriod) {
        return await this.GET(`/website/${website_id}/stats`, {
            ...(0, time_periods_1.convertPeriodToTime)(period),
        });
    }
    /**
     * Gets the pageviews of a website from a specified time period using it's ID
     * @param website_id The website's ID (not UUID)
     * @param period The time period of pageviews to return
     * @param options.unit The interval of time/precision of the returned pageviews
     * @param options.tz The timezone you're in (defaults to "America/Toronto")
     * @returns The website's pageviews from the specified time period
     * @see {@link https://github.com/umami-software/umami/blob/master/pages/api/website/[id]/pageviews.js Relevant Umami source code}
     */
    async getPageViews(website_id, period = this._defaultPeriod, options = { unit: "hour", tz: "America/Toronto" }) {
        return await this.GET(`/website/${website_id}/pageviews`, {
            ...(0, time_periods_1.convertPeriodToTime)(period),
            ...options,
        });
    }
    /**
     * Gets the events of a website from a specified time period using it's ID
     * @param website_id The website's ID (not UUID)
     * @param period The time period of events to return
     * @param options.unit The interval of time/precision of the returned pageviews
     * @param options.tz The timezone you're in (defaults to "America/Toronto")
     * @returns An array of events from the specified time period
     * @see {@link https://github.com/umami-software/umami/blob/master/pages/api/website/[id]/events.js Relevant Umami source code}
     */
    async getEvents(website_id, period = this._defaultPeriod, options = { unit: "hour", tz: "America/Toronto" }) {
        return await this.GET(`/website/${website_id}/events`, {
            ...(0, time_periods_1.convertPeriodToTime)(period),
            ...options,
        });
    }
    async getEventsBy(website_id, filter, value, period = this._defaultPeriod) {
        const events = await this.getMetrics(website_id, period, { type: "event" });
        return events.reduce((total, { x, y }) => {
            const [type, name] = x.split("\t");
            switch (filter) {
                case "type":
                    if (type == value) {
                        return total + y;
                    }
                    break;
                case "name":
                    if (name == value) {
                        return total + y;
                    }
                    break;
            }
            return total;
        }, 0);
    }
    /**
     * Gets a type of metrics of a website from a specified time period using it's ID
     * @param website_id The website's ID (not UUID)
     * @param period The time period of events to return
     * @param options.type The type of metric to get. Defaults to url
     * @returns An array of metrics from the specified time period
     * @see {@link https://github.com/umami-software/umami/blob/master/pages/api/website/[id]/metrics.js Relevant Umami source code}
     */
    async getMetrics(website_id, period = this._defaultPeriod, options = { type: "url" }) {
        return await this.GET(`/website/${website_id}/metrics`, {
            ...(0, time_periods_1.convertPeriodToTime)(period),
            ...options,
        });
    }
    async getActiveVisitors(website_id) {
        return await this.GET(`/website/${website_id}/active`);
    }
    /**
     * Gets all the user accounts
     * @returns An array of all the user accounts
     * @see {@link https://github.com/umami-software/umami/blob/master/pages/api/accounts/index.js Relevant Umami source code}
     */
    async getAccounts() {
        this.assertAdmin("You are not authorized to get the accounts.");
        return await this.GET("/accounts");
    }
}
